File: src/main/kotlin/com/linqingying/cangjie/psi/CjFunctionLiteral.java
package com.linqingying.cangjie.psi;

import com.linqingying.cangjie.lexer.CjTokens;
import com.linqingying.cangjie.name.FqName;
import com.linqingying.cangjie.name.SpecialNames;
import com.intellij.lang.ASTNode;
import com.intellij.psi.PsiElement;
import com.intellij.psi.search.LocalSearchScope;
import com.intellij.psi.search.SearchScope;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;


public class CjFunctionLiteral extends CjFunctionNotStubbed {
    public CjFunctionLiteral(@NotNull ASTNode node) {
        super(node);
    }

    @Override
    public boolean hasBlockBody() {
        return false;
    }

    @Override
    public String getName() {
        return SpecialNames.ANONYMOUS_STRING;
    }

    @Override
    public PsiElement getNameIdentifier() {
        return null;
    }

    public boolean hasParameterSpecification() {
        return findChildByType(CjTokens.ARROW) != null;
    }

    @Override
    public CjBlockExpression getBodyExpression() {
        return (CjBlockExpression) super.getBodyExpression();
    }

    @Nullable
    @Override
    public PsiElement getEqualsToken() {
        return null;
    }

    @NotNull
    public PsiElement getLBrace() {
        return findChildByType(CjTokens.LBRACE);
    }

    @Nullable
    @IfNotParsed
    public PsiElement getRBrace() {
        return findChildByType(CjTokens.RBRACE);
    }

    @Nullable
    public PsiElement getArrow() {
        return findChildByType(CjTokens.ARROW);
    }

    @Nullable
    @Override
    public FqName getFqName() {
        return null;
    }

    @Override
    public boolean hasBody() {
        return getBodyExpression() != null;
    }

    @NotNull
    @Override
    public SearchScope getUseScope() {
        return new LocalSearchScope(this);
    }
}

File: src/main/kotlin/com/linqingying/cangjie/psi/CjFunctionLiteral.kt
package com.linqingying.cangjie.psi

import com.linqingying.cangjie.lexer.CjTokens
import com.linqingying.cangjie.name.FqName
import com.linqingying.cangjie.name.SpecialNames
import com.intellij.lang.ASTNode
import com.intellij.psi.PsiElement
import com.intellij.psi.search.LocalSearchScope
import com.intellij.psi.search.SearchScope

class CjFunctionLiteral(node: ASTNode) : CjFunctionNotStubbed(node) {
    override fun hasBlockBody(): Boolean {
        return false
    }

    override fun getName(): String {
        return SpecialNames.ANONYMOUS_STRING
    }

    override fun getNameIdentifier(): PsiElement? {
        return null
    }

    fun hasParameterSpecification(): Boolean {
        return findChildByType<PsiElement?>(CjTokens.ARROW) != null
    }

    override val bodyExpression: CjBlockExpression?
        get() {
            return super.bodyExpression as CjBlockExpression?
        }


    override val equalsToken: PsiElement? = null
    val lBrace: PsiElement
        get() = findChildByType(CjTokens.LBRACE)!!

    @get:IfNotParsed
    val rBrace: PsiElement?
        get() = findChildByType(CjTokens.RBRACE)

    val arrow: PsiElement?
        get() = findChildByType(CjTokens.ARROW)

    override val fqName: FqName?
        get() = null

    override fun hasBody(): Boolean {
        return bodyExpression != null
    }

    override fun getUseScope(): SearchScope {
        return LocalSearchScope(this)
    }
}
